source/common/network/dns_impl.h

```diff
@@ -5,43 +5,60 @@
 #include "common/common/linked_object.h"
 #include "common/event/dispatcher_impl.h"
 
+#include "ares.h"
 #include "event2/event_struct.h"
 
 namespace Network {
 
+class DnsResolverImplPeer;
+
 /**
- * Implementation of DnsResolver that uses getaddrinfo_a. All calls and callbacks are assumed to
- * happen on the thread that owns the creating dispatcher. Also, since results come in via signal
- * only one of these can exist at a time.
+ * Implementation of DnsResolver that uses c-ares. All calls and callbacks are assumed to
+ * happen on the thread that owns the creating dispatcher.
  */
 class DnsResolverImpl : public DnsResolver {
 public:
   DnsResolverImpl(Event::DispatcherImpl& dispatcher);
-  ~DnsResolverImpl();
+  ~DnsResolverImpl() override;
 
   // Network::DnsResolver
-  ActiveDnsQuery& resolve(const std::string& dns_name, ResolveCb callback) override;
+  ActiveDnsQuery* resolve(const std::string& dns_name, ResolveCb callback) override;
 
 private:
-  struct PendingResolution : LinkedObject<PendingResolution>, public ActiveDnsQuery {
+  friend class DnsResolverImplPeer;
+  struct PendingResolution : public ActiveDnsQuery {
     // Network::ActiveDnsQuery
-    void cancel() override { cancelled_ = true; }
+    void cancel() override {
+      // c-ares only supports channel-wide cancellation, so we just allow the
+      // network events to continue but don't invoke the callback on completion.
+      cancelled_ = true;
+    }
+
+    // c-ares ares_gethostbyname() query callback.
+    void onAresHostCallback(int status, struct hostent* hostent);
 
-    std::string host_;
-    addrinfo hints_;
-    gaicb async_cb_data_;
+    // Caller supplied callback to invoke on query completion or error.
     ResolveCb callback_;
-    bool cancelled_{};
+    // Does the object own itself? Resource reclamation occurs via self-deleting
+    // on query completion or error.
+    bool owned_ = false;
+    // Has the query completed? Only meaningful if !owned_;
+    bool completed_ = false;
+    // Was the query cancelled via cancel()?
+    bool cancelled_ = false;
   };
 
-  typedef std::unique_ptr<PendingResolution> PendingResolutionPtr;
-
-  void onSignal();
+  // libevent callback for events on sockets tracked in events_.
+  void onEventCallback(evutil_socket_t socket, short events, void* arg);
+  // c-ares callback when a socket state changes, indicating that libevent
+  // should listen for read/write events.
+  void onAresSocketStateChange(int fd, int read, int write);
+  // Initialize the channel with given ares_init_options().
+  void initializeChannel(struct ares_options* options, int optmask);
 
   Event::DispatcherImpl& dispatcher_;
-  int signal_fd_;
-  event signal_read_event_;
-  std::list<PendingResolutionPtr> pending_resolutions_;
+  ares_channel channel_;
+  std::unordered_map<int, event*> events_;
```

Comment:
Please use dispatcher_.createFileEvent(). It will remove a lot of boilerplate and make this code not actually depend on libevent directly. Note that right now FileEvent does not allow setting a timeout. I'm not sure if it's better to modify the the FileEvent interface to add timeout capability, or to just add a parallel Event::Timer for timeouts.